# HOUSE PRICE PREDICTION

## Loading the data from sklearn library

In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import pickle

In [2]:
from sklearn.datasets import fetch_california_housing

In [3]:
data = fetch_california_housing()

## Gathering info about the data

In [4]:
print(data.DESCR)

.. _california_housing_dataset:

California Housing dataset
--------------------------

**Data Set Characteristics:**

    :Number of Instances: 20640

    :Number of Attributes: 8 numeric, predictive attributes and the target

    :Attribute Information:
        - MedInc        median income in block
        - HouseAge      median house age in block
        - AveRooms      average number of rooms
        - AveBedrms     average number of bedrooms
        - Population    block population
        - AveOccup      average house occupancy
        - Latitude      house block latitude
        - Longitude     house block longitude

    :Missing Attribute Values: None

This dataset was obtained from the StatLib repository.
http://lib.stat.cmu.edu/datasets/

The target variable is the median house value for California districts.

This dataset was derived from the 1990 U.S. census, using one row per census
block group. A block group is the smallest geographical unit for which the U.S.
Census Bur

In [5]:
df = pd.DataFrame()

In [6]:
data.data

array([[   8.3252    ,   41.        ,    6.98412698, ...,    2.55555556,
          37.88      , -122.23      ],
       [   8.3014    ,   21.        ,    6.23813708, ...,    2.10984183,
          37.86      , -122.22      ],
       [   7.2574    ,   52.        ,    8.28813559, ...,    2.80225989,
          37.85      , -122.24      ],
       ...,
       [   1.7       ,   17.        ,    5.20554273, ...,    2.3256351 ,
          39.43      , -121.22      ],
       [   1.8672    ,   18.        ,    5.32951289, ...,    2.12320917,
          39.43      , -121.32      ],
       [   2.3886    ,   16.        ,    5.25471698, ...,    2.61698113,
          39.37      , -121.24      ]])

## Merging all the independent feature into one DATAFRAME

In [7]:
data.data.shape

(20640, 8)

In [8]:
df = pd.DataFrame(data = data.data, columns = data.feature_names)
df.head()

MedInc  HouseAge  AveRooms  AveBedrms  Population  AveOccup  Latitude  \
0  8.3252      41.0  6.984127   1.023810       322.0  2.555556     37.88   
1  8.3014      21.0  6.238137   0.971880      2401.0  2.109842     37.86   
2  7.2574      52.0  8.288136   1.073446       496.0  2.802260     37.85   
3  5.6431      52.0  5.817352   1.073059       558.0  2.547945     37.85   
4  3.8462      52.0  6.281853   1.081081       565.0  2.181467     37.85   

   Longitude  
0    -122.23  
1    -122.22  
2    -122.24  
3    -122.25  
4    -122.25

## Adding the dependent features

In [9]:
df['Target'] = data.target

In [10]:
data.target

array([4.526, 3.585, 3.521, ..., 0.923, 0.847, 0.894])

In [11]:
df.head()

MedInc  HouseAge  AveRooms  AveBedrms  Population  AveOccup  Latitude  \
0  8.3252      41.0  6.984127   1.023810       322.0  2.555556     37.88   
1  8.3014      21.0  6.238137   0.971880      2401.0  2.109842     37.86   
2  7.2574      52.0  8.288136   1.073446       496.0  2.802260     37.85   
3  5.6431      52.0  5.817352   1.073059       558.0  2.547945     37.85   
4  3.8462      52.0  6.281853   1.081081       565.0  2.181467     37.85   

   Longitude  Target  
0    -122.23   4.526  
1    -122.22   3.585  
2    -122.24   3.521  
3    -122.25   3.413  
4    -122.25   3.422

## Exploratory Data Analysis

In [22]:
pip install sweetviz

     |████████████████████████████████| 15.1 MB 3.8 MB/s eta 0:00:01
Note: you may need to restart the kernel to use updated packages.


In [25]:
import sweetviz as sv

In [26]:
report = sv.analyze(df)

                                             |      | [  0%]   00:00 -> (? left)

In [27]:
report.show_html("./report.html")

Report ./report.html was generated! NOTEBOOK/COLAB USERS: the web browser MAY not pop up, regardless, the report IS saved in your notebook/colab files.


## Data Preprocessing
### We can convert Latitude and Longitude into addresses which will be much easier for us to understand

In [28]:
pip install geopy

     |████████████████████████████████| 118 kB 4.1 MB/s eta 0:00:01
Note: you may need to restart the kernel to use updated packages.


In [29]:
from geopy.geocoders import Nominatim

In [30]:
geolocator = Nominatim(user_agent = 'geoapiExercises')

In [53]:
geolocator.reverse("37.88"+" , "+"-122.23" ).raw['address']

{'leisure': 'Ecological Study Area',
 'road': 'Upper Side Hill entrance',
 'city': 'Oakland',
 'county': 'Alameda County',
 'state': 'California',
 'ISO3166-2-lvl4': 'US-CA',
 'postcode': '94563',
 'country': 'United States',
 'country_code': 'us'}

In [44]:
def location(cord):
    latitude = str(cord[0])
    longitude = str(cord[1])
    
    location = geolocator.reverse( latitude + "," + longitude).raw['address']
    
    if location.get('road') is None:
        location['road'] = None
    
    if location.get('county') is None:
        location['county'] = None
    
    loc_update['county'].append(location['county'])
    loc_update['road'].append(location['road'])

In [61]:
loc_update = {"county": [],
             "road": []}

for i,cord in enumerate(df.iloc[:,6:-1].values):
    location(cord)
    
    pickle.dump(loc_update, open('loc_update.pickle','wb'))
    
    if i % 100 == 0:
        print(i)

0
100
200
300
400
500
600
700
800
900
1000
1100
1200
1300
1400
1500
1600
1700
1800
1900
2000
2100
2200
2300
2400
2500
2600
2700
2800
2900
3000
3100
3200
3300
3400
3500
3600
3700
3800
3900
4000
4100
4200
4300
4400
4500
4600
4700
4800
4900
5000
5100
5200
5300
5400
5500
5600
5700
5800
5900
6000
6100
6200
6300
6400
6500
6600
6700
6800
6900
7000
7100
7200
7300
7400
7500
7600
7700
7800
7900
8000
8100
8200
8300
8400
8500
8600
8700
8800
8900
9000
9100
9200
9300
9400
9500
9600
9700
9800
9900
10000
10100
10200
10300
10400
10500
10600
10700
10800
10900
11000
11100
11200
11300
11400
11500
11600
11700
11800
11900
12000
12100
12200
12300
12400
12500
12600
12700
12800
12900
13000
13100
13200
13300
13400
13500
13600
13700
13800
13900
14000
14100
14200
14300
14400
14500
14600
14700
14800
14900
15000
15100
15200
15300
15400
15500
15600
15700
15800
15900
16000
16100
16200
16300
16400
16500
16600
16700
16800
16900
17000
17100
17200
17300
17400
17500
17600
17700
17800
17900
18000
18100
18200
18300
18400
18

In [12]:
loc_update = pickle.load(open("loc_update.pickle", "rb"))

In [13]:
loc_update

{'county': ['Alameda County',
  'Contra Costa County',
  'Alameda County',
  'Alameda County',
  'Alameda County',
  'Alameda County',
  'Alameda County',
  'Alameda County',
  'Alameda County',
  'Alameda County',
  'Alameda County',
  'Alameda County',
  'Alameda County',
  'Alameda County',
  'Alameda County',
  'Alameda County',
  'Alameda County',
  'Alameda County',
  'Alameda County',
  'Alameda County',
  'Alameda County',
  'Alameda County',
  'Alameda County',
  'Alameda County',
  'Alameda County',
  'Alameda County',
  'Alameda County',
  'Alameda County',
  'Alameda County',
  'Alameda County',
  'Alameda County',
  'Alameda County',
  'Alameda County',
  'Alameda County',
  'Alameda County',
  'Alameda County',
  'Alameda County',
  'Alameda County',
  'Alameda County',
  'Alameda County',
  'Alameda County',
  'Alameda County',
  'Alameda County',
  'Alameda County',
  'Alameda County',
  'Alameda County',
  'Alameda County',
  'Alameda County',
  'Alameda County',
  'Al

In [14]:
loc = pd.DataFrame(loc_update)

In [15]:
loc.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20640 entries, 0 to 20639
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   county  20068 non-null  object
 1   road    19580 non-null  object
dtypes: object(2)
memory usage: 322.6+ KB


## Add the new features to the dataframe

In [16]:
for i in loc_update.keys():
    df[i] = loc_update[i]
    
df = df.sample(axis = 0, frac = 1)
df.head(10)

MedInc  HouseAge  AveRooms  AveBedrms  Population  AveOccup  Latitude  \
18060  8.0331      34.0  7.910314   1.089686       692.0  3.103139     37.26   
2122   2.4728      18.0  5.349406   1.103038      4476.0  2.956407     36.74   
13509  2.1250      29.0  5.335714   1.021429       991.0  3.539286     34.13   
6165   4.1172      37.0  5.256614   1.055556      1279.0  3.383598     34.07   
3867   3.1478      22.0  4.165839   1.078882      2620.0  1.627329     34.17   
11941  3.5403      26.0  5.159915   1.134328      1383.0  2.948827     33.94   
11872  2.9688      42.0  4.162544   0.922261       804.0  2.840989     33.99   
9560   3.0911      27.0  5.528509   1.061404      1307.0  2.866228     37.36   
6293   3.6275      23.0  4.876344   1.026882      3881.0  4.173118     34.01   
19998  2.0313      18.0  5.440000   1.093333       253.0  3.373333     36.09   

       Longitude   Target                 county                    road  
18060    -122.02  5.00001     Santa Clara County           Douglass Lane  
2122     -119.71  0.88300          Fresno County  East Huntington Avenue  
13509    -117.34  0.70600  San Bernardino County       North Macy Street  
6165     -117.95  1.76500     Los Angeles County    South Sunkist Avenue  
3867     -118.46  2.73700     Los Angeles County           Lemona Avenue  
11941    -117.42  1.13500       Riverside County         Garfield Street  
11872    -117.36  0.92900       Riverside County       Riverside Freeway  
9560     -120.60  0.86900          Merced County           Bellevue Road  
6293     -117.89  1.54100     Los Angeles County        Willow Glen Lane  
19998    -119.45  1.12500          Tulare County              Avenue 168

### Drop latitude, longitude column

In [17]:
df = df.drop(labels = ["Latitude", "Longitude"], axis = 1)
df.head()

MedInc  HouseAge  AveRooms  AveBedrms  Population  AveOccup   Target  \
18060  8.0331      34.0  7.910314   1.089686       692.0  3.103139  5.00001   
2122   2.4728      18.0  5.349406   1.103038      4476.0  2.956407  0.88300   
13509  2.1250      29.0  5.335714   1.021429       991.0  3.539286  0.70600   
6165   4.1172      37.0  5.256614   1.055556      1279.0  3.383598  1.76500   
3867   3.1478      22.0  4.165839   1.078882      2620.0  1.627329  2.73700   

                      county                    road  
18060     Santa Clara County           Douglass Lane  
2122           Fresno County  East Huntington Avenue  
13509  San Bernardino County       North Macy Street  
6165      Los Angeles County    South Sunkist Avenue  
3867      Los Angeles County           Lemona Avenue

### Classification algorithm to find missing Categorical value

In [18]:
missing_idx = []
for i in range(df.shape[0]):
    if df['road'][i] is None:
        missing_idx.append(i)
        
missing_road_X_train = np.array([[df['MedInc'][i], df['AveBedrms'][i], df['AveRooms'][i]] for i in range(df.shape[0]) if i not in missing_idx])

missing_road_Y_train = np.array([df['road'][i] for i in range(df.shape[0]) if i not in missing_idx])

missing_road_X_test = np.array([[df['MedInc'][i], df['AveBedrms'][i], df['AveRooms'][i]] for i in range(df.shape[0]) if i in missing_idx])

In [19]:
missing_road_X_train

array([[8.3252    , 1.02380952, 6.98412698],
       [8.3014    , 0.97188049, 6.23813708],
       [7.2574    , 1.07344633, 8.28813559],
       ...,
       [1.7       , 1.12009238, 5.20554273],
       [1.8672    , 1.17191977, 5.32951289],
       [2.3886    , 1.16226415, 5.25471698]])

In [20]:
missing_road_Y_train

array(['Upper Side Hill entrance', 'Skyline Trail',
       'Grove Shafter Freeway', ..., 'Indiana Ranch Road', 'Vierra Road',
       'Marysville Road'], dtype='<U77')

In [21]:
missing_road_X_test.shape

(1060, 3)

### We Use Stochastic Gradient Classifier

In [22]:
from sklearn.linear_model import SGDClassifier

#### Model Training

In [23]:
model_1 = SGDClassifier()
model_1.fit(missing_road_X_train, missing_road_Y_train)

SGDClassifier()

In [24]:
missing_road_y_pred = model_1.predict(missing_road_X_test)

In [25]:
np.unique(missing_road_y_pred)

array(['11th Avenue', 'Alemany Boulevard', 'Bailey Drive',
       'Baronet Place', 'Bayside Walk', 'Cleveland Avenue',
       'East 106th Street', 'Front Street', 'Haynes Street',
       'Howard Road', 'Ida Lane', 'Malcolm Avenue',
       'Mount Tallac Trail (17E33B)', 'Noble Avenue', 'Pine Tree Terrace',
       'West 134th Street', 'West Washington Boulevard',
       'Whittier Boulevard'], dtype='<U77')

### Add the model back to the dataframe

In [26]:
for n, i in enumerate(missing_idx):
    df['road'][i] = missing_road_y_pred[n]

/var/folders/4f/9wwls0j92z90fgmqsv9knhsc0000gn/T/ipykernel_35547/1130247970.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['road'][i] = missing_road_y_pred[n]


In [27]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 20640 entries, 18060 to 1713
Data columns (total 9 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   MedInc      20640 non-null  float64
 1   HouseAge    20640 non-null  float64
 2   AveRooms    20640 non-null  float64
 3   AveBedrms   20640 non-null  float64
 4   Population  20640 non-null  float64
 5   AveOccup    20640 non-null  float64
 6   Target      20640 non-null  float64
 7   county      20068 non-null  object 
 8   road        20640 non-null  object 
dtypes: float64(7), object(2)
memory usage: 2.1+ MB


### Label Encode on the Categorical Data

In [28]:
from sklearn.preprocessing import LabelEncoder

In [29]:
le = LabelEncoder()
df['road'] = le.fit_transform(df['road'])

In [30]:
df.head()

MedInc  HouseAge  AveRooms  AveBedrms  Population  AveOccup   Target  \
18060  8.0331      34.0  7.910314   1.089686       692.0  3.103139  5.00001   
2122   2.4728      18.0  5.349406   1.103038      4476.0  2.956407  0.88300   
13509  2.1250      29.0  5.335714   1.021429       991.0  3.539286  0.70600   
6165   4.1172      37.0  5.256614   1.055556      1279.0  3.383598  1.76500   
3867   3.1478      22.0  4.165839   1.078882      2620.0  1.627329  2.73700   

                      county  road  
18060     Santa Clara County  2198  
2122           Fresno County  2489  
13509  San Bernardino County  5526  
6165      Los Angeles County  7582  
3867      Los Angeles County  4394

### Classification algorithm to find missing Categorical value(county) same as we did for road

In [31]:
missing_idx = []
for i in range(df.shape[0]):
    if df['county'][i] is None:
        missing_idx.append(i)
        
missing_county_X_train = np.array([[df['MedInc'][i], df['AveBedrms'][i], df['AveRooms'][i]] for i in range(df.shape[0]) if i not in missing_idx])

missing_county_Y_train = np.array([df['county'][i] for i in range(df.shape[0]) if i not in missing_idx])

missing_county_X_test = np.array([[df['MedInc'][i], df['AveBedrms'][i], df['AveRooms'][i]] for i in range(df.shape[0]) if i in missing_idx])

In [32]:
from sklearn.linear_model import SGDClassifier

In [33]:
model_1 = SGDClassifier()
model_1.fit(missing_county_X_train, missing_county_Y_train)

SGDClassifier()

In [34]:
missing_county_y_pred = model_1.predict(missing_county_X_test)

In [35]:
np.unique(missing_county_y_pred)

array(['Los Angeles County', 'Placer County', 'Santa Cruz County',
       'Trinity County'], dtype='<U22')

In [36]:
for n, i in enumerate(missing_idx):
    df['county'][i] = missing_county_y_pred[n]

/var/folders/4f/9wwls0j92z90fgmqsv9knhsc0000gn/T/ipykernel_35547/832331889.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['county'][i] = missing_county_y_pred[n]


In [37]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 20640 entries, 18060 to 1713
Data columns (total 9 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   MedInc      20640 non-null  float64
 1   HouseAge    20640 non-null  float64
 2   AveRooms    20640 non-null  float64
 3   AveBedrms   20640 non-null  float64
 4   Population  20640 non-null  float64
 5   AveOccup    20640 non-null  float64
 6   Target      20640 non-null  float64
 7   county      20640 non-null  object 
 8   road        20640 non-null  int64  
dtypes: float64(7), int64(1), object(1)
memory usage: 2.1+ MB


In [38]:
le = LabelEncoder()
df['county'] = le.fit_transform(df['county'])

In [39]:
df.head()

MedInc  HouseAge  AveRooms  AveBedrms  Population  AveOccup   Target  \
18060  8.0331      34.0  7.910314   1.089686       692.0  3.103139  5.00001   
2122   2.4728      18.0  5.349406   1.103038      4476.0  2.956407  0.88300   
13509  2.1250      29.0  5.335714   1.021429       991.0  3.539286  0.70600   
6165   4.1172      37.0  5.256614   1.055556      1279.0  3.383598  1.76500   
3867   3.1478      22.0  4.165839   1.078882      2620.0  1.627329  2.73700   

       county  road  
18060      44  2198  
2122       10  2489  
13509      38  5526  
6165       20  7582  
3867       20  4394

## Train-Test Split

In [40]:
y = df.iloc[:,-3].values
df.drop(labels= ['Target'], axis= 1, inplace = True)
x = df.iloc[:,:].values

In [41]:
from sklearn.model_selection import train_test_split

In [42]:
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.20, random_state = 42)

## Model Selection

In [48]:
from sklearn.ensemble import RandomForestRegressor

In [49]:
model = RandomForestRegressor()
model.fit(x_train, y_train)

RandomForestRegressor()

In [50]:
y_pred = model.predict(x_test)

## Checking the accuracy of the model

In [51]:
from sklearn.metrics import r2_score

accuracy = r2_score(y_test, y_pred)

In [52]:
ACCURACY = accuracy * 100
print("Accuracy of the model is: ",ACCURACY)

Accuracy of the model is:  74.70855406614727


### Adding our own test data

In [55]:
inp_data = np.array([4.1172,	37.0,	5.256614,	1.055556,	1279.0,	3.383598,	20,	7582])

In [56]:
inp = inp_data.reshape((1, -1))

In [57]:
model.predict(inp)

array([1.88299])

#### Result for this input data is 1.765 which is close to 1.883